mirakuru/base.py

```diff
@@ -18,16 +18,60 @@
 """Base executor with the most basic functionality."""
 
 import os
+import re
 import time
 import shlex
 import signal
 import subprocess
+import uuid
+import logging
 from contextlib import contextmanager
 
 from mirakuru.exceptions import TimeoutExpired, AlreadyRunning
 
 
-class Executor(object):
+log = logging.getLogger(__name__)
+
+
+class KidsMarker(object):
+
+    """
+    Helper class for a trick with marking subprocesses with env variable.
+    
+    As there is no easy way to determine all child processes that were created
+    during lifetime of a certain process we do a hack with environment
+    variable which can be added to subprocesses in order to mark who was the
+    original parent even if some process daemonized or changed process group.
+    There may be a situation that subprocess will abandon orignal envs from
+    parent so we also try to extract subprocesses names markers.
+    """
+
+    ENV_UUID = 'mirakuru_uuid'
+    PS_AUX = ('ps', 'auxe', '-ww')
+    PID_MATCH = re.compile('^\S+[ ]+(\d+).*$')
+
+    def __init__(self):
+        self._uuid = str(uuid.uuid1())
+
+    def kids_marked_env(self):
+        env = os.environ.copy()
+        env[self.__class__.ENV_UUID] = self._uuid
+        return env
+
+    def kids_kill(self, sig):
+        env = '%s=%s' % (self.ENV_UUID, self._uuid)
+
+        for line in subprocess.check_output(self.PS_AUX).splitlines():
+            if env not in line:
```

Comment:
``` python
if env not in str(line):
```

For python3 compatibility. Unless you'll find a way to return py3 str from check_output, as right now it's basestring.